File: src/main/java/com/zippeykeys/praisebe/block/multi/Multiblock.java
package com.zippeykeys.praisebe.block.multi;

import java.util.ArrayList;
import java.util.List;

import com.zippeykeys.praisebe.factory.MultiblockBuilder;
import com.zippeykeys.praisebe.iface.ILocalize;

import org.apache.logging.log4j.util.TriConsumer;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.Nullable;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.var;

import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.registries.IForgeRegistryEntry;


@AllArgsConstructor
public class Multiblock extends IForgeRegistryEntry.Impl<Multiblock> implements ILocalize{
    @Getter(onMethod_ = @Override)
    private final String name;

    @Getter
    private final EMBActivator activator;

    @Getter
    private final IBlockState trigger;

    @Getter
    public final List<MultiblockPart> parts = new ArrayList<>();

    private final TriConsumer<Multiblock, World, BlockPos> structureBuilder;

    public boolean isValid(World world, BlockPos pos, EntityPlayer player){
        for(var i = 0; i < 4; i++){
            if(checkStructure(world, pos, player)){
                return true;
            }
            getParts().forEach(MultiblockPart::rotate);
        }
        return false;
    }

    public boolean addPart(MultiblockPart part){
        if(getPartAt(part.getPos()) != null){
            return false;
        }
        return parts.add(part);
    }

    public @Nullable MultiblockPart getPartAt(BlockPos pos){
        return getParts() //
          .stream() //
          .filter(x -> x.getPos() == pos) //
          .findAny() //
          .orElse(null);
    }

    public boolean checkStructure(World world, BlockPos pos, EntityPlayer player){
        return getParts() //
          .stream() //
          .allMatch(x -> world.getBlockState(pos.add(x.getPos()))
            .equals(x.getBlockState()));
    }

    public void buildStructure(World world, BlockPos blockPos){
        structureBuilder.accept(this, world, blockPos);
    }

    @Override
    public String getPrefix(){
        return "multiblock";
    }

    @Contract("_ -> new")
    public static MultiblockBuilder builder(TriConsumer<Multiblock, World, BlockPos> structureBuilder){
        return new MultiblockBuilder(structureBuilder);
    }
}

File: src/main/kotlin/com/zippeykeys/praisebe/block/multi/Multiblock.kt
package com.zippeykeys.praisebe.block.multi

import com.google.common.collect.ImmutableList
import com.zippeykeys.praisebe.iface.ILocalize
import com.zippeykeys.praisebe.util.getResource
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraftforge.registries.IForgeRegistryEntry
import org.apache.logging.log4j.util.TriConsumer

class Multiblock(private val actualName: String, val activator: EMBActivator = EMBActivator.values()[0],
        val trigger: IBlockState, private val parts: ImmutableList<MultiblockPart>,
        private val structureBuilder: TriConsumer<Multiblock, World, BlockPos>) :
        IForgeRegistryEntry.Impl<Multiblock>(),
        ILocalize {
    init {
        registryName = getResource(actualName)
    }

    override fun getName(): String {
        return actualName
    }

    override val prefix: String
        get() = "multiblock"

    fun isValid(world: World, pos: BlockPos, player: EntityPlayer): Boolean {
        for (i in 0..3) {
            if (checkStructure(world, pos, player)) {
                return true
            }
            parts.forEach { it.rotate() }
        }
        return false
    }

    operator fun get(pos: BlockPos): MultiblockPart? {
        return parts
                .stream()
                .filter { x -> x.pos === pos }
                .findAny()
                .orElse(null)
    }

    fun checkStructure(world: World, pos: BlockPos, player: EntityPlayer): Boolean {
        return parts
                .stream()
                .allMatch { x -> world.getBlockState(pos.add(x.pos)) == x.blockState }
    }

    fun buildStructure(world: World, blockPos: BlockPos) {
        structureBuilder.accept(this, world, blockPos)
    }
}